gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/services/SubscriptionService.java

```diff
@@ -72,12 +110,94 @@ public Optional<Subscription> getById(String subscriptionId) {
-            }
-            saveInCacheBySubscriptionId(subscription);
+            cacheStandardSubscription(subscription);
         } else if (Subscription.Type.SUBSCRIPTION == subscription.getType()) {
-            subscriptionDispatcher.dispatch(subscription);
+            cacheDispatchableSubscription(subscription);
+        }
+    }
+
+    @Override
+    public void saveOrDispatch(Subscription subscription) {
+        if (Subscription.Type.STANDARD == subscription.getType()) {
+            cacheStandardSubscription(subscription);
+        } else if (Subscription.Type.SUBSCRIPTION == subscription.getType()) {
+            dispatch(subscription);
+        }
+    }
+
+    @Override
+    public void dispatchFor(List<String> apis) {
+        apis.forEach(
+            api -> {
+                final List<Subscription> subscriptions = dispatchableSubscriptionsByApi.evict(api);
+                if (subscriptions != null) {
+                    subscriptions.forEach(this::dispatch);
+                }
+            }
+        );
+    }
+
+    private void cacheStandardSubscription(Subscription subscription) {
+        if (subscription.getClientId() != null) {
+            saveInCacheByApiAndClientId(subscription);
+        }
+        saveInCacheBySubscriptionId(subscription);
+    }
+
+    private void cacheDispatchableSubscription(Subscription subscription) {
+        final List<Subscription> subscriptions = dispatchableSubscriptionsByApi.get(subscription.getApi());
+        List<Subscription> subs = subscriptions;
+        if (subscriptions == null) {
+            subs = new ArrayList<>();
+        }
+        subs.add(subscription);
+        dispatchableSubscriptionsByApi.put(subscription.getApi(), subs);
+    }
+
+    private void dispatch(Subscription subscription) {
+        subscriptionDispatcher
+            .dispatch(subscription)
+            .doOnComplete(() -> LOGGER.info("Subscription [{}] has been dispatched", subscription.getId()))
+            .onErrorResumeNext(t -> sendFailureCommand(subscription, t))
+            .subscribe();
+    }
+
+    private Completable sendFailureCommand(Subscription subscription, Throwable throwable) {
+        return Completable
+            .fromRunnable(
+                () -> {
+                    final Command command = new Command();
+                    Instant now = Instant.now();
+                    command.setId(UUID.random().toString());
+                    command.setFrom(node.id());
+                    command.setTo(MessageRecipient.MANAGEMENT_APIS.name());
+                    command.setTags(List.of(CommandTags.SUBSCRIPTION_FAILURE.name()));
+                    command.setCreatedAt(Date.from(now));
+                    command.setUpdatedAt(Date.from(now));
+
+                    convertSubscriptionCommand(subscription, command, throwable.getMessage());
+
+                    LOGGER.info("Subscription [{}] failed with cause: {}", subscription.getId(), throwable.getMessage());
```

Comment:
Shouldn't we use an error log for this case as we are in an error "path" 🤔 

```suggestion
                    LOGGER.error("Subscription [{}] failed with cause: {}", subscription.getId(), throwable.getMessage());
```

Reply:
good catch, and I move it in the `onErrorResumeNextLambda` as discussed